src/gtc/numpy/npir_codegen.py

```diff
@@ -87,189 +93,105 @@ def __setitem__(self, key, value):
-                extent + slice_to_extent(acc)
-            )
-
-
 class NpirCodegen(TemplatedGenerator):
-    def visit_DataType(self, node: common.DataType, **kwargs: Any) -> Union[str, Collection[str]]:
-        return f"np.{node.name.lower()}"
-
-    def visit_BuiltInLiteral(self, node: common.BuiltInLiteral, **kwargs) -> str:
-        if node is common.BuiltInLiteral.TRUE:
-            return "True"
-        elif node is common.BuiltInLiteral.FALSE:
-            return "False"
-        else:
-            return self.generic_visit(node, **kwargs)
-
-    Literal = FormatTemplate("{dtype}({value})")
-
-    BroadCast = FormatTemplate("{expr}")
-
-    Cast = FormatTemplate("np.array({expr}, dtype={dtype})")
-
-    def visit_NumericalOffset(
-        self, node: npir.NumericalOffset, **kwargs: Any
-    ) -> Union[str, Collection[str]]:
-        operator, delta = op_delta_from_int(node.value)
-        return self.generic_visit(node, op=operator, delta=delta, **kwargs)
-
-    NumericalOffset = FormatTemplate("{op}{delta}")
-
-    def visit_AxisOffset(self, node: npir.AxisOffset, **kwargs: Any) -> Union[str, Collection[str]]:
-        offset = self.visit(node.offset)
-        axis_name = self.visit(node.axis_name)
-        lpar, rpar = "()" if offset else ("", "")
-        variant = self.AxisOffset_parallel if node.parallel else self.AxisOffset_serial
-        rendered = variant.render(lpar=lpar, rpar=rpar, axis_name=axis_name, offset=offset)
-        return self.generic_visit(node, parallel_or_serial_variant=rendered, **kwargs)
-
-    AxisOffset_parallel = JinjaTemplate(
-        "{{ lpar }}{{ axis_name | lower }}{{ offset }}{{ rpar }}:{{ lpar }}{{ axis_name | upper }}{{ offset }}{{ rpar }}"
-    )
+    @dataclass
+    class BlockContext:
+        locals_declared: Set[str] = field(default_factory=set)
 
-    AxisOffset_serial = JinjaTemplate(
-        "{{ lpar }}{{ axis_name | lower }}_{{ offset }}{{ rpar }}:({{ axis_name | lower }}_{{ offset}} + 1)"
-    )
+        def add_declared(self, *args):
+            self.locals_declared |= set(args)
 
-    AxisOffset = FormatTemplate("{parallel_or_serial_variant}")
+    contexts = (SymbolTableTrait.symtable_merger,)
 
     def visit_FieldDecl(self, node: npir.FieldDecl, **kwargs) -> Union[str, Collection[str]]:
-        if all(node.dimensions):
-            return ""
-        shape_idx = iter(range(3))
-        origin_idx = iter(range(3))
-        shape = ", ".join(
-            [f"{node.name}.shape[{next(shape_idx)}]" if dim else "1" for dim in node.dimensions]
-        )
-        origin = ", ".join(
-            [
-                f"_origin_['{node.name}'][{next(origin_idx)}]" if dim else "0"
-                for dim in node.dimensions
-            ]
+        dims = _paren_wrap(", ".join(["True" if d else "False" for d in node.dimensions]))
+        return self.generic_visit(node, dims=dims, **kwargs)
+
+    FieldDecl = FormatTemplate("{name} = Field({name}, _origin_['{name}'], {dims})")
```

Comment:
Just a suggestion, it should be possible to inline the dims part in the template:

```suggestion
    FieldDecl = FormatTemplate("{name} = Field({name}, _origin_['{name}'], ({', '.join('True' if d else 'False' for d in dimensions)}))")
```